lib/hanami/view/helpers/escape_helper.rb

```diff
@@ -0,0 +1,301 @@
+      #
+      # By including <tt>Hanami::Helpers::EscapeHelper</tt> it will inject private
+      # methods as markup escape utilities.
+      #
+      # @since 0.1.0
+      module EscapeHelper
+        module_function
+
+        # Escape the given HTML tag content.
+        #
+        # This should be used only for untrusted contents: user input.
+        #
+        # This should be used only for tag contents.
+        # To escape tag attributes please use <tt>Hanami::Helpers::EscapeHelper#escape_html_attribute</tt>.
+        #
+        # @param input [String] the input
+        #
+        # @return [String] the escaped string
+        #
+        # @since 0.1.0
+        #
+        # @see Hanami::Helpers::EscapeHelper#escape_html_attribute
+        #
+        # @example Basic usage
+        #   require 'hanami/helpers/escape_helper'
+        #
+        #   class MyView
+        #     include Hanami::Helpers::EscapeHelper
+        #
+        #     def good_content
+        #       h "hello"
+        #     end
+        #
+        #     def evil_content
+        #       h "<script>alert('xss')</script>"
+        #     end
+        #   end
+        #
+        #   view = MyView.new
+        #
+        #   view.good_content
+        #     # => "hello"
+        #
+        #   view.evil_content
+        #     # => "&lt;script&gt;alert(&apos;xss&apos;)&lt;&#x2F;script&gt;"
+        #
+        # @example With HTML builder
+        #   #
+        #   # CONTENTS ARE AUTOMATICALLY ESCAPED
+        #   #
+        #   require 'hanami/helpers'
+        #
+        #   class MyView
+        #     include Hanami::Helpers
+        #
+        #     def evil_content
+        #       html.div do
+        #         "<script>alert('xss')</script>"
+        #       end
+        #     end
+        #   end
+        #
+        #   view = MyView.new
+        #   view.evil_content
+        #     # => "<div>\n&lt;script&gt;alert(&apos;xss&apos;)&lt;&#x2F;script&gt;</div>"
+        def escape_html(input)
+          Temple::Utils.escape_html_safe(input)
+        end
+
+        # @since 0.1.0
+        alias_method :h, :escape_html
+
+        # Returns an escaped, HTML safe string from a given array.
+        #
+        # Behaves similarly to `Array#join`. In addition, given array is flattened, and all items,
+        # including the supplied separator, are HTML escaped unless they are already HTML safe. The
+        # returned string is also marked as HTML safe.
+        #
+        # @example
+        #   safe_join([raw("<p>foo</p>"), "<p>bar</p>"], "<br />")
+        #   # => "<p>foo</p>&lt;br /&gt;&lt;p&gt;bar&lt;/p&gt;"
+        #
+        #   safe_join([raw("<p>foo</p>"), raw("<p>bar</p>")], raw("<br />"))
+        #   # => "<p>foo</p><br /><p>bar</p>"
+        #
+        # @api public
+        # @since 2.0.0
+        def escape_join(array, sep = $,)
+          sep = escape_html(sep)
+
+          array.flatten.map { |i| escape_html(i) }.join(sep).html_safe
```

Comment:
I actually tried this! But it broke some specs. I'll get back to you here with what happened.

Reply:
So here's the tests that fail when we do this:

Test:

```
expect(tag.input(value: [123, "abc", [789]])).to eq %(<input value="123 abc 789">)
```

Failure:

```
  1) Hanami::View::Helpers::TagHelper#tag tag attributes joins arrays of options and converts them to strings
     Failure/Error: expect(tag.input(value: [123, "abc", [789]])).to eq %(<input value="123 abc 789">)

       expected: "<input value=\"123 abc 789\">"
            got: "<input value=\"123 abc [789]\">"

       (compared using ==)
     # ./spec/unit/helpers/tag_helper_spec.rb:133:in `block (4 levels) in <top (required)>'
```

Test:

```
expect(escape_join(["<script>", ["<script>"]], " ")).to eq %(&lt;script&gt; &lt;script&gt;)
```

Failure:

```
  2) Hanami::View::Helpers::EscapeHelper#escape_join flattens the given array
     Failure/Error: expect(escape_join(["<script>", ["<script>"]], " ")).to eq %(&lt;script&gt; &lt;script&gt;)

       expected: "&lt;script&gt; &lt;script&gt;"
            got: "&lt;script&gt; [&quot;&lt;script&gt;&quot;]"

       (compared using ==)
     # ./spec/unit/helpers/escape_helper_spec.rb:45:in `block (3 levels) in <top (required)>'
```

So the reason for the implementation is to allow arrays in arrays to be passed to `#safe_join`, making its behaviour similar to Ruby's own `Array#join`:

```ruby
[123, "abc", [789]].join("--")
# => "123--abc--789"
```

The PR introducing this to Rails' equivalent helper is this one from June 2014: https://github.com/rails/rails/pull/15654

Given the parallel to `Array#join`, I think it'd make sense to leave this implementation as is.

I did just realise we could replace `flatten.map` with `flatten.map!` to avoid an extra array allocation, however, since the `flatten` is already giving us a new array. I'll push that up in another commit in just a moment.